wnfs/src/private/directory.rs

```diff
@@ -616,6 +616,96 @@ impl PrivateDirectory {
         }
     }
 
+    /// Opens a mutable reference to the specified file.
+    /// If the file is missing, it initializes an empty file and give a mut reference to that.
+    /// If the file already exists, it will copy it to the next revision, update the edit time, and give a mut reference to that.
+    /// # Examples
+    /// ```
+    /// use std::rc::Rc;
+    /// use chrono::Utc;
+    /// use rand::thread_rng;
+    /// use wnfs::{
+    ///    private::{PrivateForest, PrivateRef, PrivateDirectory},
+    ///    common::{BlockStore, MemoryBlockStore},
+    ///    namefilter::Namefilter,
+    /// };
+    /// #[async_std::main]
+    /// async fn main() {
+    ///    let store = &mut MemoryBlockStore::default();
+    ///    let rng = &mut thread_rng();
+    ///    let mut forest : &mut Rc<PrivateForest> = &mut Rc::new(PrivateForest::new());
+    ///    let root_dir = &mut Rc::new(PrivateDirectory::new(
+    ///         Namefilter::default(),
+    ///         Utc::now(),
+    ///         rng,
+    ///    ));
+    ///    let content = b"print('hello world')";
+    ///    root_dir
+    ///        .write(
+    ///            &["code".into(), "hello.py".into()],
+    ///            true,
+    ///            Utc::now(),
+    ///            content.to_vec(),
+    ///            &mut forest,
+    ///            store,
+    ///            rng
+    ///        )
+    ///        .await
+    ///        .unwrap();
+    ///    let mut file_forest = &mut Rc::clone(forest);
+    ///    let mut file = {
+    ///        root_dir
+    ///            .open_file_mut(&["code".into(), "hello.py".into()], true, Utc::now(), forest, store, rng)
+    ///            .await
+    ///            .unwrap()
+    ///    };
+    ///    file.set_content(
+    ///        Utc::now(),
+    ///        &b"print('hello world 2')"[..],
+    ///        file_forest,
+    ///        store,
+    ///        rng,
+    ///    );
+    ///    let forest = &forest.merge(&file_forest, store).await.unwrap();
+    ///    let result = root_dir
+    ///        .read(&["code".into(), "hello.py".into()], true, forest, store)
+    ///        .await
+    ///        .unwrap();
+    ///    assert_eq!(&result, b"print('hello world 2')");
+    /// }
+    /// ```
+    pub async fn open_file_mut<'a>(
+        self: &'a mut Rc<Self>,
+        path_segments: &[String],
+        search_latest: bool,
+        time: DateTime<Utc>,
+        forest: &'a mut Rc<PrivateForest>,
+        store: &'a mut impl BlockStore,
```

Comment:
I took another look at this today, I think this works :)

```suggestion
        forest: &mut Rc<PrivateForest>,
        store: &mut impl BlockStore,
```

No need for `forest.merge` then and we can also re-use the same BlockStore :+1: